File: Core/src/main/java/dev/lynith/core/events/EventBus.java
package dev.lynith.core.events;

import dev.lynith.core.Logger;

import java.lang.reflect.Method;
import java.util.concurrent.CopyOnWriteArrayList;

public class EventBus<CB> {

    private final Logger logger;

    public EventBus() {
        logger = new Logger("eventbus");
    }

    public EventBus(String name) {
        logger = new Logger(name);
    }

    private final CopyOnWriteArrayList<StoredEvent> events = new CopyOnWriteArrayList<>();

    public <T extends CB> void on(Class<T> event, T callback) {
        events.add(new StoredEvent(event, callback, false));
    }

    public <T extends CB> void once(Class<T> event, T callback) {
        events.add(new StoredEvent(event, callback, true));
    }

    public <T extends CB> void emit(Class<T> callbackClass, Object... args) {
        mainLoop:
        for (StoredEvent event : events) {
            try {
                Class<T> eventClass = (Class<T>) event.getEvent();
                T callback = (T) event.getCallback();

                if (eventClass.isAssignableFrom(callbackClass) && eventClass.getName().equals(callbackClass.getName())) {

                    for (Method method : callbackClass.getMethods()) {
                        if (method.getName().equals("invoke")) {
                            if (method.getParameterCount() == args.length) {
                                try {
                                    method.invoke(callback, args);
                                } catch (Exception e) {
                                    logger.error("Error while invoking event {}: {}", eventClass.getName());
                                    e.printStackTrace();
                                }
                                break;
                            }

                            logger.error("Invalid number of arguments for event " + eventClass.getName());
                            break;
                        }

                    }

                    if (event.isOnce()) {
                        events.remove(event);
                    }
                }
            } catch (Exception e) {
                logger.error("Error while emitting event {}", event.getEvent().getName());
                e.printStackTrace();
            }
        }
    }

    private static class StoredEvent<CB> {
        private final Class<? extends CB> event;
        private final CB callback;
        private final boolean once;

        public StoredEvent(Class<? extends CB> event, CB callback, boolean once) {
            this.event = event;
            this.callback = callback;
            this.once = once;
        }

        public Class<? extends CB> getEvent() {
            return event;
        }

        public CB getCallback() {
            return callback;
        }

        public boolean isOnce() {
            return once;
        }
    }

}

File: Core/src/main/kotlin/dev/lynith/core/events/EventBus.kt
package dev.lynith.core.events

import dev.lynith.core.Logger
import java.util.concurrent.CopyOnWriteArrayList

open class EventBus<E : Event> @JvmOverloads constructor(
    name: String = "eventbus"
) {

    private val logger: Logger = Logger(name)
    private val events: MutableMap<Class<out E>, MutableList<StoredEvent<in E>>> = mutableMapOf()

    fun <T : E> on(eventClazz: Class<T>, callback: (T) -> Unit, once: Boolean = false) {
        val storedEvent = StoredEvent(callback, once) as StoredEvent<in E>
        events[eventClazz] = (events[eventClazz] ?: CopyOnWriteArrayList()).apply {
            add(storedEvent)
        }
    }

    fun <T : E> once(eventClazz: Class<T>, callback: (T) -> Unit) {
        on(eventClazz, callback, true)
    }

    inline fun <reified T : E> on(noinline callback: (T) -> Unit) {
        on(T::class.java, callback)
    }

    inline fun <reified T : E> once(noinline callback: (T) -> Unit) {
        on(T::class.java, callback, true)
    }

    open fun emit(event: E) {
        val callbacks = events[event.javaClass] ?: return

        for (callback in callbacks) {
            callback.call(event)
            if (callback.once) {
                events[event.javaClass]?.remove(callback)
            }
        }
    }

    private class StoredEvent<E : Event> (
        val callback: (E) -> Unit,
        var once: Boolean = false
    ) {
        fun call(event: E) {
            callback.invoke(event)
        }

        override fun toString(): String {
            return "StoredEvent(callback=${callback.javaClass.simpleName}, once=$once)"
        }
    }

}
